### PR TITLE
Enhance diagram with full supertypes ancestry relationship, change arrowheads and add guards to avoid errors

### PR DESCRIPTION
**Pull request by BarkingBad**
As I rewrited pretty much previous PR and some comments became irrelevant, I decided to close previous PR and open new one to get fresh view on changes. However, I'd like to address two issues that still remain valid:
1. @abgruszecki, I assume your [comment](https://github.com/lampepfl/scala3doc/pull/212#discussion_r510744601) wanted me to try something like this:
`val parents = if smbl.exists then hackGetParents(classDef) else None` instead of `val parents = if smbl.exists then hackGetParents(smbl.tree.asInstanceOf[ClassDef]) else None`, am I correct? I tried to shortened that, and suprisingly for me these lines are not equal. I cannot guarantee the first diagram with my first approach is correct, but the second one is certainly not (example for `scala3doc/example/level2/Documentation`)

Zrzut ekranu z 2020-10-23 18-06-10
Zrzut ekranu z 2020-10-23 18-07-50

2. @TheElectronWill, you said that CSS should be changed. I was trying to represent background colours to much these from the old scaladoc. These colours were matched with white font, so I followed that convenction either. However, it is nice to have links that actually can be navigable embedded into diagram. It would be hard for user to distinguish which parts of diagram are links, and which one are plain text. Could you propose style that you see would fit best, so we can discuss it?

### CODE REVIEW

File: src/main/scala/dotty/dokka/model/api/api.scala [hunk 1]
```diff
@@ -99,9 +99,28 @@ extension (s: Signature):
 
 case class LinkToType(signature: Signature, dri: DRI, kind: Kind)
 
-case class HierarchyDiagram(edges: Seq[Edge])
-case class Vertex(val id: Int, val body: LinkToType)
-case class Edge(val from: Vertex, val to: Vertex)
+case class Vertex(id: Int, body: LinkToType)
+case class Edge(from: Vertex, to: Vertex)
+case class HierarchyGraph private (edges: Seq[Edge], private val nextId: Int):
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/model/api/api.scala]: Maybe we can store only Edge data (`(LinkToType, LinkToType)`) in the graph and compute Vertexes and Edges on demand? Then we do not need thsi complicated logic to store nextId and code will be much faster

File: src/main/scala/dotty/dokka/model/api/api.scala [hunk 2]
```diff
@@ -99,9 +99,28 @@ extension (s: Signature):
...
+    val newNextId = if higherId >= nextId then higherId + 1 else nextId 
+    HierarchyGraph(edges ++ Seq(Edge(from, to)), newNextId)
+
+  def ++(edges: Seq[(LinkToType, LinkToType)]) = edges.foldLeft(this) {
+    case (acc, edge) => acc + edge
+  }
+
+object HierarchyGraph:
+  def empty = HierarchyGraph(Seq.empty, 0)
+  def withEdge(edge: (LinkToType, LinkToType)) = HierarchyGraph.empty + edge
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/model/api/api.scala]: those methods seems to be unused

File: src/main/scala/dotty/dokka/model/api/api.scala (hunk 1 above)
Comment (romanowski) [reply src/main/scala/dotty/dokka/model/api/api.scala]: TBH we can store graph as just pairs and use that pairs to generate graph itself so probably we may not need at all:
```
case class Vertex(id: Int, body: LinkToType)
case class Edge(from: Vertex, to: Vertex)
```

File: src/main/scala/dotty/dokka/translators/ScalaPageCreator.scala [hunk 3]
```diff
@@ -426,13 +427,15 @@ class ScalaPageCreator(
...
-            if subtypes.isEmpty && supertypes.isEmpty then withSubtypes else
-                withSubtypes.header(2, "Type hierarchy")().group(
-                    kind = ContentKind.Comment,
-                    styles = Set(ContentStyle.WithExtraAttributes), 
-                    extra = PropertyContainer.Companion.empty plus SimpleAttr.Companion.header("Type hierarchy")
-                ) { _.group(kind = ContentKind.Symbol, styles = Set(TextStyle.Monospace)) { 
-                        _.dotDiagram(HierarchyDiagramBuilder.build(c))
+            
+            graph.fold(withSubtypes) { g =>
+                if g.edges.size == 0 then withSubtypes else
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/translators/ScalaPageCreator.scala]: `g.edges.isEmpty`

Comment (romanowski) [reply src/main/scala/dotty/dokka/translators/ScalaPageCreator.scala]: We should probably also rename `g` to `graph`

**Review by romanowski — APPROVED** [review]
Thank you for another iteration of this feature and I think we are almost there.

I suggested how it can be simplyfied even more :)

**Comment by abgruszecki** [comment]
@BarkingBad OK, I'm adding the issue with the `ClassDef -> symbol -> tree` route to the litany of issues we need to open in Dotty repo post-merge.

**Comment by TheElectronWill** [comment]
>I was trying to represent background colours to much these from the old scaladoc. These colours were matched with white font, so I followed that convenction either. However, it is nice to have links that actually can be navigable embedded into diagram. It would be hard for user to distinguish which parts of diagram are links, and which one are plain text. Could you propose style that you see would fit best, so we can discuss it?

@BarkingBad I see two options:
1. Keep the background colors, the white text, and simply underline resolved links like you did. Also add a "hint" on hover to confirm that "yes, this is a link":
```css
.diagram-class a:hover {
    color: #BFE7F3;
}
```
2. Use a white background, which allows links to be colored in blue like outside of the diagram.

I like the background colors, actually :slightly_smiling_face: They provide valuable and immediate information. Since it's quite intuitive that types in the diagram are clickable (when the link is resolved, at least), I think that underline+hint is enough. Therefore I would prefer option 1.

File: src/main/scala/dotty/dokka/model/api/api.scala [hunk 4]
```diff
@@ -98,10 +98,16 @@ extension (s: Signature):
...
 case class LinkToType(signature: Signature, dri: DRI, kind: Kind)
-
-case class HierarchyDiagram(edges: Seq[Edge])
-case class Vertex(val id: Int, val body: LinkToType)
-case class Edge(val from: Vertex, val to: Vertex)
+case class HierarchyGraph(edges: Seq[(LinkToType, LinkToType)]):
+  def vertecies: Seq[LinkToType] = edges.flatten((a, b) => Seq(a, b)).distinct
+  val verteciesWithId: Map[LinkToType, Int] = vertecies.zipWithIndex.toMap
+  def +(edge: (LinkToType, LinkToType)): HierarchyGraph = HierarchyGraph((edges :+ edge).distinct)
+  def ++(edges: Seq[(LinkToType, LinkToType)]): HierarchyGraph = edges.foldLeft(this) {
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/model/api/api.scala]: Maybe:
```
def +(newEdge: (LinkToType, LinkToType)): HierarchyGraph = this ++ Seq(newEdge)
def ++(newEdge: Seq[(LinkToType, LinkToType)]): HierarchyGraph = HierarchyGraph((edges ++ newEdge).distinct)
```

File: src/main/scala/dotty/dokka/model/api/api.scala [hunk 5]
```diff
@@ -98,10 +98,16 @@ extension (s: Signature):
   def join(a: Signature): Signature = s ++ a
 
 case class LinkToType(signature: Signature, dri: DRI, kind: Kind)
-
-case class HierarchyDiagram(edges: Seq[Edge])
-case class Vertex(val id: Int, val body: LinkToType)
-case class Edge(val from: Vertex, val to: Vertex)
+case class HierarchyGraph(edges: Seq[(LinkToType, LinkToType)]):
+  def vertecies: Seq[LinkToType] = edges.flatten((a, b) => Seq(a, b)).distinct
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/model/api/api.scala:102]: Usually `flatMap` is used in such cases, I am supprised that `flatten` works

File: src/main/scala/dotty/dokka/model/api/api.scala [hunk 6]
```diff
@@ -98,10 +98,16 @@ extension (s: Signature):
...
+  val verteciesWithId: Map[LinkToType, Int] = vertecies.zipWithIndex.toMap
```
Comment (romanowski) [inline src/main/scala/dotty/dokka/model/api/api.scala]: imho it should be lazy

**Review by romanowski — COMMENTED** [review]
Bunch of comments realted to reduce unnecessary object alocation in Graph implemetation.

File: src/main/scala/dotty/dokka/model/api/api.scala (hunk 6 above)
Comment (BarkingBad) [reply src/main/scala/dotty/dokka/model/api/api.scala]: Yeah, good point

File: src/main/scala/dotty/dokka/model/api/api.scala (hunk 5 above)
Comment (BarkingBad) [reply src/main/scala/dotty/dokka/model/api/api.scala:102]: Would you like me to change it then?

File: src/main/scala/dotty/dokka/model/api/api.scala (hunk 4 above)
Comment (BarkingBad) [reply src/main/scala/dotty/dokka/model/api/api.scala]: Can change

**Comment by BarkingBad** [comment]
Thank you for your response @TheElectronWill. I also like option 1, so I will add your CSS snippet

File: src/main/scala/dotty/dokka/model/api/api.scala (hunk 5 above)
Comment (romanowski) [reply src/main/scala/dotty/dokka/model/api/api.scala:102]: we can leave it for now

**Pull request merged**